Log top origin of opener window when tracking cross-tab WindowProxy usage

bnham opened the pull request:
#### 3d18e91f254e56164b990747f07c3a6c471f785a
<pre>
Log top origin of opener window when tracking cross-tab WindowProxy usage
<a href="https://bugs.webkit.org/show_bug.cgi?id=274999">https://bugs.webkit.org/show_bug.cgi?id=274999</a>
<a href="https://rdar.apple.com/129085170">rdar://129085170</a>

Reviewed by John Wilander.

When tracking cross-tab WindowProxy usage, currently we are logging the site of the opener frame
(which could be an iframe) and the site of the openee frame. Instead we want to log the site
associated with the opener&apos;s main frame, since what we care about is potential sharing of top-level
cookies between the tabs.

Additionally, we should suppress logging when there is cross-tab WindowProxy usage but the two tab&apos;s
main frames have the same site. The two tabs are already sharing top-level cookies in that case so
we don&apos;t need to log in that case.

Finally, we used to make the distinction between logging the actual property type being accessed
(postMessage vs. closed vs. other). But due to event logging constraints we probably don&apos;t want to
be doing this for now as it is tripling the event volume in many cases. So for now we just log every
property access in the &quot;other&quot;&quot; category.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/opener/cross-site-child-access-parent-iframe-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/cross-site-child-access-parent-iframe-via-windowproxy.html: Added.
* LayoutTests/http/wpt/opener/resources/iframe-open-window.html: Added.
* LayoutTests/http/wpt/opener/resources/send-message-to-opener.html:
* LayoutTests/http/wpt/opener/same-site-child-access-parent-iframe-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/same-site-child-access-parent-iframe-via-windowproxy.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::didAccessWindowProxyPropertyViaOpener):

Canonical link: <a href="https://commits.webkit.org/279783@main">https://commits.webkit.org/279783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a584e4126684e69beb322c635ac171f001b05a9f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54531 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33948 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7100 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57809 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5263 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/56834 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5266 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44160 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3530 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56627 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32090 "Found 60 new test failures: accessibility/accessibility-node-reparent.html accessibility/ancestor-computation.html accessibility/announcement-notification.html accessibility/aria-braillelabel.html accessibility/aria-brailleroledescription.html accessibility/aria-checked-mixed-value.html accessibility/aria-current-state-changed-notification.html accessibility/aria-current.html accessibility/aria-describedby-on-input.html accessibility/aria-description.html ... (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47224 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25295 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28887 "Found 60 new test failures: imported/w3c/web-platform-tests/FileAPI/Blob-methods-from-detached-frame.html imported/w3c/web-platform-tests/FileAPI/FileReader/Progress_event_bubbles_cancelable.html imported/w3c/web-platform-tests/FileAPI/FileReader/workers.html imported/w3c/web-platform-tests/FileAPI/FileReaderSync.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-array-buffer.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-bytes.any.worker.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-dom.window.html imported/w3c/web-platform-tests/FileAPI/blob/Blob-constructor-endings.html ... (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4554 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3404 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/50663 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4768 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59400 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29763 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/4927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51584 "Found 1 new test failure: imported/w3c/web-platform-tests/pointerevents/pointerevent_pointer_boundary_events_after_reappending_last_over_target.html (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30911 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47311 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50954 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/memory-pressure (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11918 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/31893 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30701 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

johnwilander commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1295,17 +1295,29 @@ void LocalFrame::didAccessWindowProxyPropertyViaOpener(WindowProxyProperty prope
+    if (!opener())
+        return;
+
+    auto openerFrame = dynamicDowncast<LocalFrame>(opener()->mainFrame());
```
I think `openerMainFrame` is a better name for this variable.

johnwilander commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1295,17 +1295,29 @@ void LocalFrame::didAccessWindowProxyPropertyViaOpener(WindowProxyProperty prope
+    if (!openerFrame)
+        return;
+
+    auto parentTopOrigin = SecurityOriginData::fromFrame(openerFrame);
```
Where does the term `parent` come in here? Aren't we just getting the origin of `openerMainFrame`? If so, I think this should be called `openerMainFrameOrigin`.

johnwilander commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1295,17 +1295,29 @@ void LocalFrame::didAccessWindowProxyPropertyViaOpener(WindowProxyProperty prope
-    auto parentOrigin = SecurityOriginData::fromURL(parentWindow->location().url());
-    if (parentOrigin.isNull() || parentOrigin.isOpaque())
+    auto site = RegistrableDomain(origin);
+    auto parentTopSite = RegistrableDomain(parentTopOrigin);
```
Then this becomes `openerMainFrameSite` or `openerMainFrameRegistrableDomain`.

johnwilander commented on Source/WebCore/page/LocalFrame.cpp:
```diff
@@ -1295,17 +1295,29 @@ void LocalFrame::didAccessWindowProxyPropertyViaOpener(WindowProxyProperty prope
-    if (parentOrigin.isNull() || parentOrigin.isOpaque())
+    auto site = RegistrableDomain(origin);
+    auto parentTopSite = RegistrableDomain(parentTopOrigin);
+    if (site == parentTopSite)
```
Good.

johnwilander reviewed (approved):
Please make the requested name changes. Other than that, looks good to me.

bnham committed:
Log top origin of opener window when tracking cross-tab WindowProxy usage
https://bugs.webkit.org/show_bug.cgi?id=274999
rdar://129085170

Reviewed by John Wilander.

When tracking cross-tab WindowProxy usage, currently we are logging the site of the opener frame
(which could be an iframe) and the site of the openee frame. Instead we want to log the site
associated with the opener's main frame, since what we care about is potential sharing of top-level
cookies between the tabs.

Additionally, we should suppress logging when there is cross-tab WindowProxy usage but the two tab's
main frames have the same site. The two tabs are already sharing top-level cookies in that case so
we don't need to log in that case.

Finally, we used to make the distinction between logging the actual property type being accessed
(postMessage vs. closed vs. other). But due to event logging constraints we probably don't want to
be doing this for now as it is tripling the event volume in many cases. So for now we just log every
property access in the "other"" category.

* LayoutTests/TestExpectations:
* LayoutTests/http/wpt/opener/cross-site-child-access-parent-iframe-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/cross-site-child-access-parent-iframe-via-windowproxy.html: Added.
* LayoutTests/http/wpt/opener/resources/iframe-open-window.html: Added.
* LayoutTests/http/wpt/opener/resources/send-message-to-opener.html:
* LayoutTests/http/wpt/opener/same-site-child-access-parent-iframe-via-windowproxy-expected.txt: Added.
* LayoutTests/http/wpt/opener/same-site-child-access-parent-iframe-via-windowproxy.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-site-isolation/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
* Source/WebCore/page/LocalFrame.cpp:
(WebCore::LocalFrame::didAccessWindowProxyPropertyViaOpener):

Canonical link: https://commits.webkit.org/279783@main
